### ISSUE TITLE
Support HTTP/3 clients

### ISSUE THREAD

**Issue by abursavich** [projectcontour/contour#2582]
Add support for terminating [HTTP/3](https://quicwg.org/base-drafts/draft-ietf-quic-http.html) connections from clients.

Supporting HTTP/3 backends is out of scope for this issue.

Search Keywords: H3, HTTP3, QUIC

**Comment by youngnick** [comment]
Thanks for this issue @abursavich.

Since Contour is a control plane for Envoy, we are waiting on Envoy support for HTTP/3 (which I believe is tracked in envoyproxy/envoy#2557).

I'll mark this as blocked for now.

**Comment by moderation** [comment]
Envoy issue to add QUIC / HTTP3 docs - https://github.com/envoyproxy/envoy/issues/12923

**Comment by PKizzle** [comment]
Envoy has added documentation on how to enable QUIC / HTTP3: https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/http/http3.html?highlight=quic

**Comment by youngnick** [comment]
Thanks @PKizzle, we'll keep an eye on the blocking issue mentioned in the above page (envoyproxy/envoy#13115).

**Comment by Jean-Daniel** [comment]
Looks like ([envoyproxy/envoy#13115](https://github.com/envoyproxy/envoy/issues/13115)) has been merged.

**Comment by youngnick** [comment]
Okay, sadly this is now blocked behind #4075 as well. Putting the `blocked/needs-envoy` label back.

**Comment by buroa** [comment]
#4075 was merged @youngnick

**Comment by m-yosefpor** [comment]
any updates on this? Are there any WIP or milestone set for QUIC support?

**Comment by moderation** [comment]
Downstream HTTP/3 now considered GA in Envoy 1.22.0

https://www.envoyproxy.io/docs/envoy/v1.22.0/version_history/current#new-features

> http3: downstream HTTP/3 support is now GA! Upstream HTTP/3 also GA for specific deployments. See [here](https://www.envoyproxy.io/docs/envoy/v1.22.0/intro/arch_overview/http/http3#arch-overview-http3) for details.

**Comment by m-yosefpor** [comment]
Is this planned for v1.24?

**Comment by Nefelim4ag** [comment]
Looks like most of resources now are focused around gateway API.
And http3 mostly ignored because its a cool feature, but anyone can live & work without it for now.
(I'm personally of course want http3 just for fun and look around to the code to make it works, but without success)

On paper it is enough to just define http3 listener in contour, and use same configuration about upsteram clusters like http1/2 does.

Main solution for deploying contour in most setups was using Cloud LB, and AFAIK currently that setup doesn't support publish 443 udp/tcp at same time - so another barrier for wide availability of http3.

And even, if that possible to use contour without that with come magic (https://github.com/projectcontour/contour/issues/2539#issuecomment-1112248604) it is still not supported and not expected way to work with contour, so it also will not adopted in near time.

So, IMHO, http3 will not be implemented in contour in near time without some PR from community, because of issue with cloud LB setup, and because of more interests around supporting new K8s API like gateway.

**Comment by Jean-Daniel** [comment]
> Main solution for deploying contour in most setups was using Cloud LB, and AFAIK currently that setup doesn't support publish 443 udp/tcp at same time - so another barrier for wide availability of http3.

Since Kubernetes 1.24, the `MixedProtocolLBService` feature gate is enabled by default, so using a single LB service for TCP/UDP works (I'm already using it for DNS).

**Comment by Quentin-M** [comment]
fwiw Emissary has been supporting it since v3.0.0 of last summer (https://www.getambassador.io/docs/emissary/latest/topics/running/http3)

**Comment by m-yosefpor** [comment]
Are there any plans to support http3 in next contour release since there is priority label set on this feature?

**Comment by modzilla99** [comment]
I'd love to see it as well!

**Comment by Yuiie** [comment]
hello, any new on http3 support ?

**Comment by Legion2** [comment]
AWS just launched [QUIC support for their NLB](https://aws.amazon.com/blogs/networking-and-content-delivery/introducing-quic-protocol-support-for-network-load-balancer-accelerating-mobile-first-applications/), now only support from contour is missing to support QUIC for ingress in our application.
